### ISSUE TITLE
Docker inspect returns empty 'Parent' field for image that should have parent

### ISSUE THREAD

**Issue by dkjer** [moby/moby#19650]
I noticed during integration test errors in #16609 that 'docker inspect busybox' may report an empty Parent field.  This seems to be incorrect behavior (and causes integration tests to fail):

Repro:

```
DOCKER_HOST=tcp://0.0.0.0:2375 bundles/1.10.0-dev/binary/docker pull busybox
Using default tag: latest
latest: Pulling from library/busybox
eeee0535bf3c: Pull complete 
a3ed95caeb02: Pull complete 
Digest: sha256:c1bc9b4bffe665bf014a305cc6cf3bca0e6effeb69d681d7a208ce741dad58e0
Status: Downloaded newer image for busybox:latest

DOCKER_HOST=tcp://0.0.0.0:2375 bundles/1.10.0-dev/binary/docker inspect -f '{{.Parent}}' busybox
(empty reponse)
```

Daemon log during pull:

```
DEBU[0201] Calling POST /v1.23/images/create
DEBU[0201] POST /v1.23/images/create?fromImage=busybox%3Alatest
DEBU[0201] Trying to pull busybox from https://registry-1.docker.io v2
DEBU[0202] Increasing token expiration to: 0 seconds    
DEBU[0203] Pulling ref from V2 registry: busybox:latest 
DEBU[0203] pulling blob "sha256:eeee0535bf3cec7a24bff2c6e97481afa3d37e2cdeff277c57cb5cbdb2fa9e92" 
DEBU[0203] pulling blob "sha256:a3ed95caeb02ffe68cdd9fd84406680ae93d633cb16422d00e8a7c22955b46d4" 
DEBU[0203] Downloaded a3ed95caeb02 to tempfile /tmp/dockerlib/tmp/GetImageBlob481827976 
DEBU[0203] Downloaded eeee0535bf3c to tempfile /tmp/dockerlib/tmp/GetImageBlob511138887 
DEBU[0203] Applied tar sha256:2c84284818d186d88a16ac7fa731d4b71ba69ecfe11b4ce00413366833cb2403 to a19fcf37210d6e4e6f460897bf3f03c4cf2e467b356d218949117502d4ee36b2, size: 1113554 
DEBU[0203] Applied tar sha256:5f70bf18a086007016e948b04aed3b82103a36bea41755b6cddfaf10ace3c6ef to 1268fb74d315934dcb9a31ff1dd1549638ba79b5c1e3b21a34b8f7c341d1a44d, size: 0 
```

Daemon log during inspect:

```
DEBU[0249] Calling GET /v1.23/containers/busybox/json
DEBU[0249] GET /v1.23/containers/busybox/json
ERRO[0249] Handler for GET /v1.23/containers/busybox/json returned error: No such container: busybox
DEBU[0249] Calling GET /v1.23/images/busybox/json
DEBU[0249] GET /v1.23/images/busybox/json
```

Recommended additional integration test (this currently fails, in master):

```
diff --git a/integration-cli/docker_cli_inspect_test.go b/integration-cli/docker_cli_inspect_test.go
index c39b2a0..7c7e07d 100644
--- a/integration-cli/docker_cli_inspect_test.go
+++ b/integration-cli/docker_cli_inspect_test.go
@@ -415,3 +415,14 @@ func (s *DockerSuite) TestInspectContainerNetworkCustom(c *check.C) {
        c.Assert(err, checker.IsNil)
        c.Assert(strings.TrimSpace(out), checker.Equals, strings.TrimSpace(netOut))
 }
+
+func (*DockerSuite) TestInspectParentImage(c *check.C) {
+       testRequires(c, DaemonIsLinux)
+
+       _, _, err := dockerCmdWithError("pull", "busybox")
+       c.Assert(err, check.IsNil)
+
+       parent, err := inspectField("busybox", "Parent")
+       c.Assert(err, check.IsNil)
+       c.Assert(parent, checker.Not(checker.Equals), "")
+}
```

Version:

```
DOCKER_HOST=tcp://0.0.0.0:2375 bundles/1.10.0-dev/binary/docker version
Client:
 Version:      1.10.0-dev
 API version:  1.23
 Go version:   go1.5.3
 Git commit:   9600991-unsupported
 Built:        Mon Jan 25 03:00:52 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.10.0-dev
 API version:  1.23
 Go version:   go1.5.3
 Git commit:   9600991-unsupported
 Built:        Mon Jan 25 03:00:52 2016
 OS/Arch:      linux/amd64
```

Info:

```
DOCKER_HOST=tcp://0.0.0.0:2375 bundles/1.10.0-dev/binary/docker info
Containers: 0
 Running: 0
 Paused: 0
 Stopped: 0
Images: 1
Server Version: 1.10.0-dev
Storage Driver: aufs
 Root Dir: /tmp/dockerlib/aufs
 Backing Filesystem: extfs
 Dirs: 2
 Dirperm1 Supported: true
Execution Driver: native-0.2
Logging Driver: json-file
Plugins: 
 Volume: local
 Network: bridge null host
Kernel Version: 3.16.0-4-amd64
Operating System: Debian GNU/Linux 8 (jessie)
OSType: linux
Architecture: x86_64
CPUs: 16
Total Memory: 22.64 GiB
Name: don
ID: BTA7:EUOC:F2TE:IKBJ:A4DN:V2KV:L4ON:ZQF5:FSTL:QUVX:2GTS:TE57
Debug mode (server): true
 File Descriptors: 12
 Goroutines: 38
 System Time: 2016-01-25T03:07:47.981436541Z
 EventsListeners: 0
 Init SHA1: 
 Init Path: /ngi-persist/don/docker/bundles/1.10.0-dev/binary/docker
 Docker Root Dir: /tmp/dockerlib
Username: dkjer
Registry: https://index.docker.io/v1/
WARNING: No memory limit support
WARNING: No swap limit support
WARNING: No oom kill disable support
WARNING: No cpu cfs quota support
WARNING: No cpu cfs period support
```

**Comment by HackToday** [comment]
I tried this

```
ubuntu@speedwork:~$ docker pull busybox
Using default tag: latest
latest: Pulling from library/busybox
eeee0535bf3c: Pull complete 
a3ed95caeb02: Pull complete 
Digest: sha256:c1bc9b4bffe665bf014a305cc6cf3bca0e6effeb69d681d7a208ce741dad58e0
Status: Downloaded newer image for busybox:latest
ubuntu@speedwork:~$ docker images
REPOSITORY          TAG                 IMAGE ID            CREATED             SIZE
busybox             latest              0cb40641836c        9 days ago          1.114 MB
ubuntu@speedwork:~$ docker inspect busybox | grep -i parent
        "Parent": "",
ubuntu@speedwork:~$ docker images -a
REPOSITORY          TAG                 IMAGE ID            CREATED             SIZE
busybox             latest              0cb40641836c        9 days ago          1.114 MB
ubuntu@speedwork:~$ 
```

It seems busybox is the only image, not have other images in `docker images -a` output.  So no parent is understood.

@dkjer  what do you think busybox need to have one parent ? did you have extra images layer ?

**Comment by dkjer** [comment]
@HackToday This may well just be my misunderstanding of the busybox image, however it would still mean updating integration tests to match expectations.

My understanding from seeing [TestRmiParentImageFail](https://github.com/docker/docker/blob/master/integration-cli/docker_cli_rmi_test.go#L314) fail in #16609 is that 'busybox' is expected to have a non-empty Parent returned in order to pass integration tests.

**Comment by thaJeztah** [comment]
Hm, perhaps this is because of the new content-addressable store, where images no longer need a "parent" image to chain together all image-layer (in the new structure, an image has a list of "blobs" / "layers"; no need to traverse parent images). Perhaps in the "old" situation, busy box had an "empty" (from scratch) image as parent. Freshly pulled images no longer have this.

/cc @tonistiigi

**Comment by thaJeztah** [comment]
TL;DR an image pulled with 1.9.1, and *migrated* to 1.10 could stil have a parent (I think) but images pulled with 1.10 *don't*; see https://github.com/docker/docker/issues/19623#issuecomment-174263193

**Comment by tonistiigi** [comment]
@dkjer The difference is that you are deleting `busybox` frozen image in your test https://github.com/docker/docker/pull/16609/files#diff-4b1e56bb77ac16f2ccf956fc24cf0a82R431 (probably not a good idea), and then this image needs to be repulled and doesn't have a parent any more. I agree that this isn't quite ideal and may easily break in the future. We should probably switch the frozen images to the new content-addressable format and build an image for detecting the parent deletion.

**Comment by sam-thibault** [comment]
This is an old issue.  I will close it as stale.
